tokio/src/runtime/task/harness.rs

```diff
@@ -301,6 +303,25 @@ where
             }));
         }
 
+        if !snapshot.is_join_waker_set() {
```

Comment:
This logic is too confusing. Normally the snapshot is the value *before* the atomic operation, but here it's the value after the operation. Can we change `transition_to_join_handle_dropped` to instead return a struct with two booleans for whether it should drop the output or the waker?